contracts/Tokenomics.sol

```diff
@@ -1313,55 +1325,52 @@ contract Tokenomics is TokenomicsConstants {
-        uint256 oldInflationUntilEpochEnd = curInflationPerSecond * secondsInYearUntilEndTime;
-        // Calculate updated inflation per second
-        curInflationPerSecond = getInflationForYear(currentYear) / ONE_YEAR;
-        // Get new inflation from beginning of the year until end of last epoch
-        uint256 newInflationUntilEpochEnd = curInflationPerSecond * secondsInYearUntilEndTime;
-        // Check for inflation reduction
-        if (newInflationUntilEpochEnd > oldInflationUntilEpochEnd) {
-            revert Overflow(newInflationUntilEpochEnd, oldInflationUntilEpochEnd);
+        // Find if this epoch is the year change epoch
+        uint256 lastEpochYear = (endTime - timeLaunch) / ONE_YEAR;
+        if (numYears > lastEpochYear) {
+            revert Overflow(numYears, lastEpochYear);
+        }
+
+        // Reset staking inflation
+        // Assume there are no more than 100 epochs to clear retained staking amount
+        for (uint256 i = 0; i < 100; ++i) {
+            // This is a low level call since it must never revert
+            (bool success, ) = dispenser.call(abi.encodeCall(IDispenser.retain, ()));
+
+            // Break when all the retains are executed
+            if (!success) {
+                break;
+            }
         }
+        mapEpochStakingPoints[curEpochCounter].stakingIncentive = 0;
 
-        // Calculate inflation reduction
-        uint256 reductionInflation = oldInflationUntilEpochEnd - newInflationUntilEpochEnd;
-        // Check for effective bond cut in favor of inflation difference
-        if (reductionInflation > curEffectiveBond) {
-            revert Overflow(reductionInflation, curEffectiveBond);
+        // Update top-up fractions
+        uint256 sumTopUpFractions = _maxBondFraction + _topUpComponentFraction + _topUpAgentFraction + _stakingFraction;
+        if (sumTopUpFractions > 100) {
+            revert WrongAmount(sumTopUpFractions, 100);
         }
 
-        // Update effective bond
-        curEffectiveBond -= reductionInflation;
+        // All the adjustments will be accounted for in this epoch
+        TokenomicsPoint storage tp = mapEpochTokenomics[curEpochCounter];
 
-        // Recalculate maxBond
-        curMaxBond = (curEpochLen * curInflationPerSecond * mapEpochTokenomics[curEpochCounter].epochPoint.maxBondFraction) / 100;
-        // Adjust effective bond with a new maxBond value issued for the on-going epoch
-        curEffectiveBond += curMaxBond;
+        tp.epochPoint.maxBondFraction = uint8(_maxBondFraction);
+        tp.unitPoints[0].topUpUnitFraction = uint8(_topUpComponentFraction);
+        tp.unitPoints[1].topUpUnitFraction = uint8(_topUpAgentFraction);
+        mapEpochStakingPoints[curEpochCounter].stakingFraction = uint8(_stakingFraction);
+
+        // Calculate updated inflation per second
+        uint256 curInflationPerSecond = getInflationForYear(currentYear) / ONE_YEAR;
```

Comment:
should we use getActualInflationForYear?

Reply:
No, the `getInfkationForYear()` has correct inflation starting from a current year.

Reply:
yeah, but then we broke all the points for year 2. Can we have the old inflation in getInflationForYear(2) and use getActualInflationForYear(2) here?